File: C#/Arrays/ArrayedSet/ArrayedSet.cs
/*******************************************************
 *  ArrayedSet.cs
 *  Created by Stephen Hall on 11/17/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  Arrayed Set implementation in C#
 ********************************************************/
using System;
using System.Linq;

namespace DataStructures.Arrays.ArrayedSet
{
    /// <summary>
    /// ArrayedSet Class
    /// </summary>
    /// <typeparam name="T">Generic Type</typeparam>
    public class ArrayedSet<T>
    {
        private T[] _array;
        private int _count;
        private int _size;
        public int Count => _count;

        /// <summary>
        /// Default constructor
        /// </summary>
        public ArrayedSet()
        {
            _count = 0;
            _array = new T[_size = 4];
        }

        /// <summary>
        /// Constructor Initialized to given size
        /// </summary>
        /// <param name="size">Size to initialize array to</param>
        public ArrayedSet(int size)
        {
            _count = 0;
            _size = size > 0 ? size : 4;
            _array = new T[_size];
        }
        /// <summary>
        /// Doubles the size of the internal array
        /// </summary>
        private void Resize()
        {
            _size *= 2;
            T[] tmp = new T[_size];
            Array.Copy(_array, tmp, _count);
            _array = tmp;
        }

        /// <summary>
        /// Adds new item into the array
        /// </summary>
        /// <param name="data">Data to add into the arrayo</param>
        /// <returns>Data added into the array</returns>
        public T Add(T data)
        {
            if (data == null)
                return default(T);

            if (Contains(data))
                return default(T);

            if (_count == _size)
                Resize();

            _array[_count] = data;
            _count++;
            return data;
        }

        /// <summary>
        /// Appends the contents of an array to the ArrayedSet
        /// </summary>
        /// <param name="data">array to append</param>
        /// <returns>success|fail</returns>
        public bool Append(T[] data)
        {
            if (data == null)
                return false;
            foreach(T aData in data)
            {
                if (aData != null)
                    Add(aData);
            }
            return true;
        }

        /// <summary>
        /// Sets the data at the given index
        /// </summary>
        /// <param name="index">index to set</param>
        /// <param name="data">data to set index to</param>
        /// <returns>success|fail</returns>
        public bool Set(int index, T data)
        {
            if (!Contains(data) && (index >= 0 && index < _size))
            {
                _array[index] = data;
                return true;
            }
            return false;
        }

        /// <summary>
        /// Gets the data at the arrays given index
        /// </summary>
        /// <param name="index">Index to get data at</param>
        /// <returns>Data at the given index or default(T)</returns>
        public T Get(int index)
        {
            return index >= 0 && index < _size ? _array[index] : default(T);
        }

        /// <summary>
        /// Removes the data at arrays given index
        /// </summary>
        /// <param name="index">Index to remove</param>
        /// <returns> Data removed from the array or default(T)</returns>
        public T Remove(int index)
        {
            if (index < 0 || index > _count)
                return default(T);

            T tmp = _array[index];
            _array[index] = default(T);
            _count--;
            return tmp;
        }

        /// <summary>
        /// Resets the internal array to default size with no data
        /// </summary>
        public void Reset()
        {
            _count = 0;
            _size = 4;
            _array = new T[_size];
        }

        /// <summary>
        /// Clears all data in the array leaving size intact
        /// </summary>
        public void Clear()
        {
            for (int i = 0; i < _count; i++)
            {
                _array[i] = default(T);
            }
            _count = 0;
        }

        /// <summary>
        /// Tests to see if the data exist in the list
        /// </summary>
        /// <param name="data">data to find</param>
        /// <returns>success|fail</returns>
        private bool Contains(T data) => _array.Contains(data);
    }
}

File: PHP/Arrays/ArrayedSet/ArrayedSet.php
<?php
/*******************************************************
 *  ArrayedSet.php
 *  Created by Stephen Hall on 4/19/18.
 *  Copyright (c) 2018 Stephen Hall. All rights reserved.
 *  Arrayed Set implementation in php
 ********************************************************/

/**
 * ArrayedSet Class
 * @param <T> Generic type
 */
public class ArrayedSet {

    /**
     * Private Members
     */
    private $array = [];
    private $count = 0;
    private $size = 0;

     /**
     * ArrayedSet constructor initialized to a specific size
     * @param $size : Size to initialize the array to
     */
	public function __construct($size){
        $this->count = 0;
        if($size > 0)
            $this->size = $size;
        else
            $this->size = 4;
        $this->array = [$this->size];
    }

    /**
     * Doubles the size of the internal array
     */
	private function resize(){
        $this->size *= 2;
        $tmp = [$this->size];
        for($i = 0; $i < $this->count; $i++){
            $tmp[$i] = $this->array[$i];
        };
        $this->array = $tmp;
    }

    /**
     * Adds new item into the array
     * @param $data : Data to add into the array
     * @return mixed : data added into the array
     */
    public function add($data){
        if(!isset($data))
            return null;

        if($this->contains($data))
            return null;

        if($this->count == $this->size)
            $this->resize();

        $this->array[$this->count] = $data;
        $this->count++;
        return $data;
    }

    /**
     * Appends the contents of an array to the ArrayedSet
     * @param $data : array to append
     * @return boolean : success|fail
     */
    public function append($data){
        if(isset($data)){
            foreach ($data as $aData) {
                if (isset($aData))
                    $this->add($aData);
            }
            return true;
        }
        return false;
    }

    /**
     * Sets the data at the given index
     * @param $index : index to set
     * @param $data : data to set index to
     * @return boolean : success|fail
     */
    public function set($index, $data){
        if(!$this->contains($data)){
            if($index >= 0 && $index < $this->size) {
                $this->array[$index] = $data;
                return true;
            }
        }
        return false;
    }

    /**
     * Gets the data at the arrays given index
     * @param $index : Index to get data at
     * @return mixed : data at the given index or null
     */
    public function get($index){
        if($index >= 0 && $index < $this->size)
            return $this->array[$index];
        return null;
    }

    /**
     * Removes the data at arrays given index
     * @param $index : Index to remove
     * @return mixed : data removed from the array or null
     */
    public function remove($index){
        if ($index <= 0 || $index > $this->count || $this->count == 0)
            return null;

        $tmp = $this->array[$index];
        $this->array[$index] = null;
        $this->count--;
        return $tmp;
    }

    /**
     * Resets the internal array to default size with no data
     */
	public function reset(){
        $this->count = 0;
        $this->size = 4;
        $this->array = [$this->size];
    }

    /**
     * Clears all data in the array leaving size intact
     */
    public function clear(){
        for($i = 0; $i < $this->count; $i++){
        $this->array[$i] = null;
    }
        $this->count = 0;
    }


    /**
     * Tests to see if the data exist in the list
     * @param $data : data to find
     * @return bool : success|fail
     */
    private function contains($data){
        for($i = 0; $i < $this->size; $i++){
            if($this->array[$i] == $data)
            return true;
        }
        return false;
    }

    /**
     * Gets the current count of the array
     * @return Number of items in the array
     */
    public function count(){
        return $this->count;
    }

}